### PR TITLE
Lease limit enforcement

### PR DESCRIPTION
**Pull request by jakecoll**
Tested on dev. Re-works some logic in enforcement.

### CODE REVIEW

File: blazar/manager/enforcement.py [hunk 1]
```diff
@@ -153,6 +144,35 @@ def remove_lease_exception(self, user_name):
     def setup_usage_enforcement(self, project_name):
         self.initialize_project_allocation(project_name)
 
+    def check_user_max_leases(self, lease_values, resource_type):
+        user_leases = [
+            l for l in db_api.lease_get_all_by_user(lease_values['user_id'])
+            if l['status'] in [status.lease.ACTIVE, status.lease.PENDING]]
+
+        leases_by_resource_type = [
```
Comment (diurnalist) [inline blazar/manager/enforcement.py]: The name `_by_` in a variable immediately makes me think it's going to be a key/value map ("leases by ID"... "leases by name"). Maybe `leases_matching_resource_type`?

File: blazar/manager/enforcement.py [hunk 2]
```diff
@@ -153,6 +144,35 @@ def remove_lease_exception(self, user_name):
...
+            l['id'] for l in user_leases
+            if any([
+                x['resource_type'] == resource_type for x
+                in l['reservations']])
+        ]
+
+        existing = len(leases_by_resource_type)
+        max = CONF.enforcement.default_max_leases_by_resource_type.get(
```
Comment (diurnalist) [inline blazar/manager/enforcement.py]: Is `max` a reserved word? I don't know how Python handles that. It seems to still be ok in Python3. Maybe `max_leases` is better.

File: blazar/manager/enforcement.py [hunk 3]
```diff
@@ -153,6 +144,35 @@ def remove_lease_exception(self, user_name):
...
+            if l['status'] in [status.lease.ACTIVE, status.lease.PENDING]]
+
+        leases_by_resource_type = [
+            l['id'] for l in user_leases
+            if any([
+                x['resource_type'] == resource_type for x
+                in l['reservations']])
+        ]
+
+        existing = len(leases_by_resource_type)
```
Comment (diurnalist) [inline blazar/manager/enforcement.py]: Nit: `num_existing`

File: blazar/manager/enforcement.py [hunk 4]
```diff
@@ -153,6 +144,35 @@ def remove_lease_exception(self, user_name):
...
+        max = CONF.enforcement.default_max_leases_by_resource_type.get(
+            resource_type, -1)
+        exempt_project = self._is_exempt_project(lease_values['project_id'])
```
Comment (diurnalist) [inline blazar/manager/enforcement.py]: Nit: `is_exempt_project`

File: blazar/manager/enforcement.py [hunk 5]
```diff
@@ -153,6 +144,35 @@ def remove_lease_exception(self, user_name):
...
+
+        if (existing >= max) and (max != -1) and not exempt_project:
```
Comment (diurnalist) [inline blazar/manager/enforcement.py]: SUUUUPER nitpicky, but with the above changes, this reads like:

```python
if (num_existing >= max_leases) and (max_leases != -1) and not is_exempt_project:
```

... which reads a bit more clearly (to me.)

File: blazar/manager/enforcement.py [hunk 6]
```diff
@@ -153,6 +144,35 @@ def remove_lease_exception(self, user_name):
...
+            user_name = self._get_user_name(lease_values['user_id'])
+            raise common_ex.NotAuthorized(
+                'Requested lease contains a %s reservations, and user %s '
```
Comment (diurnalist) [inline blazar/manager/enforcement.py]: Nice message :+1:

File: blazar/manager/enforcement.py [hunk 7]
```diff
@@ -195,28 +210,23 @@ def check_lease_duration(self, lease_values, lease=None):
...
-        elif project_enforcement_id in self.project_max_lease_durations:
-            project_max_lease_duration = self.project_max_lease_durations.get(
-                project_enforcement_id)
-            if project_max_lease_duration != -1:
-                if (lease_duration) > project_max_lease_duration:
+        lease_violates_max_duration = (
+            CONF.enforcement.default_max_lease_duration != -1 and
+            lease_duration > CONF.enforcement.default_max_lease_duration)
+
+        if (lease_violates_max_duration and
```
Comment (diurnalist) [inline blazar/manager/enforcement.py]: I think some early returns can help clean up this function further.

```python

if not lease_violates_max_duration:
    return True

if self.is_exempt_project(lease_values['project_id']):
    return True

user_name = self._get_user_name(lease_values['user_id'])
lease_exception = self.get_lease_exception(user_name)

if not lease_exception:
    raise common_ex.NotAuthorized('Lease is longer than...')

if lease_duration > int(lease_exception):
    raise common_ex.NotAuthorized('Requested lease to last...')

return True
```

This makes it have much less nesting, and maybe is easier to read? I find early returns a very nice tool for long term code quality, but there is usually balance needed. This chain of checks seems like a nice opportunity though.

File: blazar/tests/manager/test_enforcement.py [hunk 8]
```diff
@@ -99,9 +99,9 @@ def test_new_lease_against_default_limit(self):
     def test_new_lease_against_project_limits(self):
         self.cfg.CONF.set_override('default_max_lease_duration', 86400,
                                    group='enforcement')
-        self.enforcement.project_max_lease_durations = {
-            'project_bar': 86400 * 2,
-        }
+        self.exempt_project = self.patch(self.enforcement,
```
Comment (diurnalist) [inline blazar/tests/manager/test_enforcement.py]: What if you just use `set_override` and add the project as exempt? Then you are also testing that the conf mechanism works.

File: blazar/manager/enforcement.py (hunk 1 above)
Comment (jakecoll) [reply blazar/manager/enforcement.py]: done

File: blazar/manager/enforcement.py (hunk 2 above)
Comment (jakecoll) [reply blazar/manager/enforcement.py]: No, max() is a built-in function so the code will still work, but it's better practice to use max_leases so I changed.

File: blazar/manager/enforcement.py (hunk 7 above)
Comment (jakecoll) [reply blazar/manager/enforcement.py]: Yea, this looks cleaner. Updated.

File: blazar/tests/manager/test_enforcement.py (hunk 8 above)
Comment (jakecoll) [reply blazar/tests/manager/test_enforcement.py]: Yep yep. I've updated and add comments in this so it's more clear what is being tested.

**Pull request closed without merging**